hostsfile/utils.go

```diff
@@ -36,8 +41,56 @@ func newHostlist(data []byte) *hostlist {
 	return &hostlist
 }
 
+func (h *hostname) Equal(hostnamev *hostname) bool {
+	if (h.wildcard != hostnamev.wildcard || h.ipv6 != hostnamev.ipv6) {
+		return false
+	}
+	if (!h.ip.Equal(hostnamev.ip)) {
+		return false
+	}
+	if (h.domain != hostnamev.domain) {
+		return false
+	}
+	return true
+}
+
+// return first match
+func (h *hostlist) FindHost(name string) (addr net.IP) {
+	var ips []net.IP;
+	ips = h.FindHosts(name)
+	if len(ips) > 0 {
+		addr = ips[0];
+	}
+	return
+}
+
+// return exact matches, if existing -> else, return wildcard
+func (h *hostlist) FindHosts(name string) (addrs []net.IP) {
+	for _, hostname := range *h {
+		if hostname.wildcard == false && hostname.domain == name {
+			addrs = append(addrs, hostname.ip)
+		}
+	}
+
+	if len(addrs) == 0 {
+		for _, hostname := range *h {
+			if hostname.wildcard == true && len(hostname.domain) < len(name) {
+				if name[len(name)-len(hostname.domain):] == hostname.domain {
+					var left string;
+					left = name[0:len(name)-len(hostname.domain)]
+					if !strings.Contains(left, ".") {
+						addrs = append(addrs, hostname.ip)
+					}
+				}
+			}
+		}
+	}
+
```

Comment:
I used a slightly different concat strategy (strings.Join). Also fixed add() with using host.Equal + added tests to check that error is thrown where it should be.